### PR TITLE
Framework for Approval protocol

### PR DESCRIPTION
**Pull request by birchmd**
## Overview
This is meant to implement the "bootstrap" half of the protocol described [here](https://rchain.atlassian.net/wiki/spaces/CORE/pages/485556483/Initializing+the+Blockchain+--+Protocol+for+generating+the+Genesis+block). The "Validator" half as well as integration into the node will come in future PRs.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-487

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
+package coop.rchain.casper.util.comm
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: It's occurring to me now that this is not the right package for this and `coop.rchain.casper.genesis` probably makes more sense.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+class ApproveBlockProtocol[
+    F[_]: Capture: Monad: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler] private (
+    val candidate: ApprovedBlockCandidate,
+    val start: Long,
+    val duration: Long) {
+  private implicit val logSource: LogSource = LogSource(this.getClass)
+
+  private val u                         = UnapprovedBlock(Some(candidate), start, duration)
+  private val serializedUnapprovedBlock = u.toByteString
+  private val candidateHash             = candidate.block.fold("")(b => PrettyPrinter.buildString(b.blockHash))
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Shouldn't we just a throw an error if the candidate's blockHash doesn't exist?

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  def run[F[_]: Monad: Capture: Time: ApproveBlockProtocol](waitTime: Long = 5000L): F[Unit] =
+    for {
+      t             <- Time[F].currentMillis
+      maybeApproved <- ApproveBlockProtocol[F].approvedBlock
+      _ <- maybeApproved match {
+            case Some(approvedBlock)
+                if (ApproveBlockProtocol[F].candidate.requiredSigs == 0 || t >= ApproveBlockProtocol[
+                  F].start + ApproveBlockProtocol[F].duration) =>
+              ApproveBlockProtocol[F].sendApprovedBlock(approvedBlock)
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Can you explain the reasoning behind `t >= ApproveBlockProtocol[F].start + ApproveBlockProtocol[F].duration`?

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  def sendUnapprovedBlock: F[Unit] =
+    for {
+      _ <- Log[F].info(s"APPROVAL: Beginning send of UnapprovedBlock $candidateHash to peers...")
+      _ <- CommUtil.sendToPeers[F](transport.UnapprovedBlock, serializedUnapprovedBlock)
+      _ <- Log[F].info(s"APPROVAL: Sent UnapprovedBlock $candidateHash to peers.")
+    } yield ()
+
+  def sendApprovedBlock(a: ApprovedBlock): F[Unit] = {
+    val serializedApprovedBlock = a.toByteString
+    val hash                    = a.candidate.flatMap(_.block).fold("")(b => PrettyPrinter.buildString(b.blockHash))
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Should we not just log a warning and then return if the blockhash doesn't exist?

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  private val sigData                   = Blake2b256.hash(candidate.toByteArray)
+
+  private val sigsF = new AtomicMonadState[F, Set[Signature]](AtomicAny(Set.empty[Signature]))
+
+  def addApproval(a: BlockApproval): F[Unit] = {
+    val validSig = for {
+      c   <- a.candidate if c == this.candidate
+      sig <- a.sig if Validate.signature(sigData, sig)
+    } yield sig
+    val sender = a.sig.fold("")(sig => Base16.encode(sig.publicKey.toByteArray))
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: "" could be more descriptive. Maybe "Empty signature"

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 1 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Hmm but we run this protocol occasionally as the chain progresses right?

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Yeah, true. This also gives a mechanism for creating signed checkpoints in the future.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 3 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: We want to wait for the duration to elapse before creating the approved block. This gives an opportunity to collect signatures in excess of the minimum. Say for example there are 10 validators and we only need signatures from 8 to make an approved block. If we get the first 8 signatures within 5 seconds, we could make an approved block right then, but suppose waiting 5 more seconds would give us the remaining 2. Setting a duration of 10-15 seconds would give us all 10 signatures and still not be a terribly long time to wait for the approved block to be created.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: So if it hasn't passed that duration you want to call `              ApproveBlockProtocol[F].sendUnapprovedBlock *> IOUtil.sleep(waitTime) *> run[F](
                waitTime)` in an else case like below?

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: That is exactly what gets called. `case _` will match both `None` and `Some(_)` when the `if t >= ...` returns false.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+import coop.rchain.comm.transport.TransportLayer
+import coop.rchain.shared._
+import coop.rchain.catscontrib.{Capture, IOUtil}
+
+import monix.execution.atomic.AtomicAny
+
+import scala.util.Try
+
+class ApproveBlockProtocol[
+    F[_]: Capture: Sync: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler] private (
```
Comment (goral09) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: `Capture` and `Sync` are the same things.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    val block: BlockMessage,
+    val requiredSigs: Int,
+    val start: Long,
+    val duration: Long,
+    private val sigsF: Ref[F, Set[Signature]]) {
+  private implicit val logSource: LogSource = LogSource(this.getClass)
+
+  val candidate                         = ApprovedBlockCandidate(Some(block), requiredSigs)
```
Comment (goral09) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: If you add `[(scalapb.field).no_box = true]` to field in proto file you won't have them wrapped in `Option`.

Comment (goral09) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: If you want I can add it as a part of the PR.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      block: BlockMessage,
+      requiredSigs: Int,
+      duration: Long): F[ApproveBlockProtocol[F]] =
+    for {
+      now   <- Time[F].currentMillis
+      sigsF <- Ref.of[F, Set[Signature]](Set.empty[Signature])
+    } yield new ApproveBlockProtocol(block, requiredSigs, now, duration, sigsF)
+
+  def blockApprovalPacketHandler[F[_]: Monad: ApproveBlockProtocol](
+      peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
```
Comment (goral09) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: `peer` is not used. It seems like an oversight?

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+      t             <- Time[F].currentMillis
+      maybeApproved <- ApproveBlockProtocol[F].approvedBlock
+      _ <- maybeApproved match {
+            case Some(_)
+                if (ApproveBlockProtocol[F].candidate.requiredSigs == 0 || t >= ApproveBlockProtocol[
+                  F].start + ApproveBlockProtocol[F].duration) =>
+              ApproveBlockProtocol[F].sendApprovedBlock
+
+            case _ =>
+              ApproveBlockProtocol[F].sendUnapprovedBlock *> IOUtil.sleep(waitTime) *> run[F](
```
Comment (goral09) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Hmm, I think `IOUtil.sleep(…)` is not a proper way to do sleep. IMO we should use [`cats-effect.Timer#sleep`](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Timer.scala#L154) and [the implementation](https://github.com/typelevel/cats-effect/blob/master/core/jvm/src/main/scala/cats/effect/internals/IOTimer.scala#L47).

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+          } yield none[Packet]
+      }
+
+  def run[F[_]: Monad: Capture: Time: ApproveBlockProtocol](waitTime: Long = 5000L): F[Unit] =
+    for {
+      t             <- Time[F].currentMillis
+      maybeApproved <- ApproveBlockProtocol[F].approvedBlock
+      _ <- maybeApproved match {
+            case Some(_)
+                if (ApproveBlockProtocol[F].candidate.requiredSigs == 0 || t >= ApproveBlockProtocol[
```
Comment (goral09) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Is `ApproveBlockProtocol[F].candidate.requiredSigs == 0` some special case?

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Yes. I had the same question originally  too. Probably means we should add a comment saying how it is unnecessary to wait for sigs if we don't require any in the first place.

Comment (goral09) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: @KentShikama What would be an example of the block that requires no signatures?

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: It's for testing. The `requiredSigs == 0` case will function identically to how the node behaves now where it just gets to create the approved block immediately and send it out. In production `requiredSigs` will certainly be positive.

Comment (goral09) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: The other thing is - I thought that this protocol is only for genesis block?

Comment (goral09) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: The other thing is - I thought that this protocol is only for genesis block?

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 9 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Make a PR to change it? I used what was already in `shared` but don't really have strong opinions on this.

Comment (goral09) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: I will, I am just making notes for both you and myself for the future.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 8 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Nope. Not all `PacketHandler`s need to know the peer that the packet came from, but some do, e.g. https://github.com/rchain/rchain/blob/dev/casper/src/main/scala/coop/rchain/casper/util/comm/BlockApproverProtocol.scala#L50
For this reason the effect constructor includes the peer https://github.com/rchain/rchain/blob/dev/comm/src/main/scala/coop/rchain/p2p/effects/PacketHandler.scala#L25

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 7 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Oh, neat. Yeah, that sounds useful.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 8 above)
Comment (goral09) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: But you don't use `peer` here and don't make a use of https://github.com/rchain/rchain/blob/dev/comm/src/main/scala/coop/rchain/p2p/effects/PacketHandler.scala#L25

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 6 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Ah I see, `Sync.delay` is the same as `Capture.capture`. Cool, feel free to remove usage of `Capture` in favour of `Sync` then.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 10 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: It is also for checkpoints along the way under assumption of weak subjectivity. See https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/

File: casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala (hunk 8 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala]: Well I hadn't yet, but was planning on it. If you will take a different approach then you can drop the `peer` parameter

**Pull request merged**